### ISSUE TITLE
How to extract ΔΔG values

### ISSUE THREAD

**Issue by xing-he529** [luo-group/SPURS#1]
Hi,

First, thank you for creating this valuable tool! As someone new to protein stability prediction, I'm trying to understand how to interpret the model outputs.

```
from spurs.inference import get_SPURS, parse_pdb
# ~ 10s
model, cfg = get_SPURS('./data/checkpoints/spurs')
pdb_name = 'DOCK1_MOUSE'
pdb_path = './data/inference_example/' + pdb_name + '.pdb'
chain = 'A'
pdb = parse_pdb(pdb_path, pdb_name, chain, cfg)
# ~ 1s
result = model(pdb,return_logist=True)
```

While studying the provided code, I noticed the model generates a tensor with 20 logits per position. Could you kindly explain how to convert 'result' to physically meaningful ΔΔG values in a CSV file (e.g., in kcal/mol)?
I apologize if this question is basic - I'm still learning about it. Any guidance would be greatly appreciated. Thank you for your patience and for making this project accessible!

Best regards,
Xing HE

**Comment by anar-rzayev** [comment]
Hi Xing,

Great question! I ran into the same puzzle while examining how SPURS produces an \(L\times 20\) tensor. Essentially:

**Per-position logits**  
SPURS returns a tensor \(\phi\) of shape \((L, 20)\), where \(L\) is the protein length. For each position \(i\) (0 to \(L-1\)) and each possible amino acid \(a\) (indexed in alphabetical order: A, C, D, E, …, Y), \(\phi[i,a]\) corresponds to the model’s predicted \(\Delta G\) if that position were replaced by amino acid \(a\). 

**Converting logits to \(\Delta\Delta G\)
   If the wild‐type (WT) residue at position \(i\) is, for example, leucine (L), you find leucine’s index in the 20‐AA dimension (call it `wt_idx`) and compute:
   \[
     \Delta\Delta G = \phi[i, \text{mutant\_idx}] 
                    \;-\; \phi[i, \text{wt\_idx}],
   \]
   where `mutant_idx` is the integer index for your new amino acid. This difference approximates how much the free energy changes (in kcal/mol) when you swap from WT to the new residue. Positive values suggest destabilization, negative values suggest stabilization.

**Practical example**  
   - Suppose position \(i\) is leucine (L) in the wild type. Let’s say leucine’s index is 9 (0-based indexing).  
   - If you’re checking a mutation to alanine (index 0), do:
     ```python
     ddG = phi[i, 0] - phi[i, 9]
     ```
     That’s your predicted \(\Delta\Delta G\).

**Exporting to CSV**  
   If you want to map these results into a CSV, you can do something like this:

   ```python
   import csv

   amino_acids = ['A','C','D','E','F','G','H','I','K','L',
                  'M','N','P','Q','R','S','T','V','W','Y']

   with open("spurs_ddG.csv", "w", newline="") as outfile:
       writer = csv.writer(outfile)
       writer.writerow(["Position", "WT_AA", "Mut_AA", "ddG_kcal_per_mol"])

       for i in range(L):
           wt_aa = wt_seq[i]  # the actual letter
           wt_idx = amino_acids.index(wt_aa)
           for m_idx, m_aa in enumerate(amino_acids):
               if m_aa == wt_aa:
                   continue
               ddG = phi[i, m_idx].item() - phi[i, wt_idx].item()
               writer.writerow([i+1, wt_aa, m_aa, ddG])
   ```

   - `wt_seq[i]` is whichever letter is at position \(i\) in your wild‐type sequence.  
   - Substituting from that letter to `m_aa` yields \(\Delta\Delta G\).  
   - The “+1” in the row means you’ll print residue numbers starting at 1 instead of 0.  

That’s all! You’ll get one CSV row per position–mutation pair, with each predicted \(\Delta\Delta G\) in kcal/mol. Let me know if this clarifies things or if you have any follow‐ups!

**Comment by xing-he529** [comment]
Thank you for the clear solution! The code worked perfectly on my test cases, and the ΔΔG CSV outputs align with expectations. Your guidance saved me hours of confusion—greatly appreciated!

**Comment by anar-rzayev** [comment]
Happy to help :)
